### PR TITLE
KAFKA-9051: Prematurely complete source offset read requests for stopped tasks

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-9051)

The changes here cause source offset readers to forcefully close when tasks fail to shut down within the graceful shutdown timeout period. When this happens, all pending and future offset read requests will throw an exception.

This is in line with the [API for the OffsetStorageReader class](https://kafka.apache.org/23/javadoc/org/apache/kafka/connect/storage/OffsetStorageReader.html#offsets-java.util.Collection-), which states that "The only case when an exception will be thrown is if the entire request failed, e.g. because the underlying storage was unavailable.". If a task is blocked on reading offsets from Kafka to the point where it has failed to shut down within the graceful shutdown timeout period, it's safe to say that the offset read request has failed and as a result, throw an exception.

Initially, I considered just returning null values from the offset reader once closed; however, this may cause source tasks to mutate some external state as if there were no offset for the requested source partitions and may negatively impact other tasks that have since been started by this connector. An exception is safer, and with the distinction between throwing one when a task has exceeded its graceful shutdown period vs. just being scheduled to stop, should not damage the functionality of tasks running in a healthy environment.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java
Comment (liukrimhrim) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java:311]: With this block of change, do we still need `secondGetInvokedAndPassed` above on line 313? It seems useless now.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java
Comment (liukrimhrim) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java:220]: Do we still need `getInvokedAndPassed`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java
Comment (liukrimhrim) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java:246]: Do we need `secondGetInvokedAndPassed`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (wicknicks) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java]: I think we should try to retain using the interface everywhere. extend the interface if you have to.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java
Comment (wicknicks) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java]: I think this would return stale values for a key. if this future was prematurely closed, then we might still return a stale values for a key. we should be careful that this doesn't break any of the existing contracts of OffsetReader.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetReadFuture.java
Comment (wicknicks) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetReadFuture.java]: can we call this method `stop()` or `forceStop()`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java
Comment (wicknicks) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java]: we should cache the fact that this future has been force stopped, and if so, any subsequent calls to get() should immediately return with either en empty map. right now, it looks like it will re-try to catch up with the remote log.

**Review by wicknicks — COMMENTED** [review]
thanks for the fix, @C0urante. some comments.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetReadFuture.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetReadFuture.java]: How about `forceComplete`? `stop` is a little ambiguous w/r/t whether threads blocked on `get` encounter a `CancellationException` or continue normally.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java]: Yes, this is noted in the description. However, if we want to stick strictly to the API for the `OffsetStorageReader`, once option we have is to simply return an empty map to all callers blocked on the future.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java]: I agree with the gist of this comment, but I think the right place for that caching behavior is in the offset reader class, not in the backing store (since each task is given its own offset reader, but the backing store is shared among all of them).

**Comment by C0urante** [comment]
Thanks for the review, @liukrimhrim and @wicknicks! I've incorporated most of your comments and responded to the rest; ready for another round when you have time

**Review by liukrimhrim — APPROVED** [review]
LGTM

**Review by wicknicks — APPROVED** [review]
LGTM. Thanks for the fix, @C0urante.

**Comment by C0urante** [comment]
@wicknicks I've altered the functionality here to be a little more permissive for tasks in a healthy Kafka cluster and (hopefully) a little less likely to provide bad data to zombie tasks that may negatively impact running (or to-be-run) tasks. Would you mind doing another round when you have time?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java
Comment (ncliang) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java]: Do we have tests to test the cancellation logic?

**Review by ncliang — COMMENTED** [review]
Looks good in general. Thanks for implementing the changes we discussed yesterday! Just one comment about testing the change.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java]: Added those now.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (ncliang) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java]: You can use ExecutorService API and replace calls to this method with just `executorService.submit()`

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (ncliang) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:168]: Shouldn't the `get()` be canceled and throw `CancellationException` above? When will the code below be executed?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java
Comment (ncliang) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java]: This implementation doesn't strictly adhere to the Future#cancel() contract - https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#cancel-boolean-

`After this method returns, subsequent calls to isDone() will always return true`. Before, we were ignoring the `mayInterruptIfRunning` flag and always cancelling when `cancel()` is called. I think I'd prefer that behavior than not actually cancelling when `cancel(false)` is called.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (ncliang) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:74]: fail() if no exception?

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (ncliang) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:91]: ditto, missing fail() call

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (ncliang) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:142]: ditto, fail()

**Review by ncliang — COMMENTED** [review]
Looks great! Just a few more comments.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:168]: It shouldn't be executed, but just in case something goes wrong, it seems more appropriate to throw that exception than just to fail the test because the `CancellationException` wasn't thrown from `get`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java]: > Before, we were ignoring the mayInterruptIfRunning flag and always cancelling when cancel() is called

I don't believe this is correct. Before, `cancel` was just a no-op and _no_ cancellation occurred, regardless of the value of `mayInterruptIfRunning`. That also violated the contract of the `Future` interface; I don't see this as a huge issue since we know that no other parts of the code base rely on the `cancel` method anyways.
Still, after reading the javadocs more carefully, I think it should be possible to implement `cancel` correctly if `mayInterruptIfRunning` is `false`.

**Comment by C0urante** [comment]
@ncliang thanks, ready for the next round

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java
Comment (ncliang) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java]: I was referring to a previous snapshot where you didn't check the flag.

**Review by ncliang — APPROVED** [review]
LGTM! Looks great!

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:74]: Ack, addressed

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:91]: Ack, addressed

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/ConvertingFutureCallbackTest.java:142]: Ack, addressed

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java]: We should first check to see if this is closed, and if so simply return.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java]: It is now possible with your changes for `ConvertingFutureCallback` to throw an exception during `cancel(boolean)`. If that happens, then this reader instance will not be properly/completely closed, so that needs to be handled.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java
Comment (rhauch) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java:313]: The bad thing about removing the callback parameter is that these tests no longer have a good way of known whether the reader completed successfully. Do we really not need this anymore?

**Review by rhauch — COMMENTED** [review]
Thanks, @C0urante. I have a few suggestions, and one question about removing the callback parameter.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java]: Ack, will address.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java]: Ack, will address.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java:313]: I could be wrong, but isn't it enough to just verify that the call to `ConvertingFutureCallback::get` returned without throwing an exception? Any additional testing logic contained in the callback (such as in https://github.com/apache/kafka/pull/7532/files#diff-d363e0391be734af7583fd6c3a30d77eL256-L260) is retained by examining the offsets returned from that call (such as in https://github.com/apache/kafka/pull/7532/files#diff-d363e0391be734af7583fd6c3a30d77eR247-R248).

Retaining this kind of API (even if it's completely internal) just for testing purposes doesn't seem like a great idea; I'd prefer to remove it if it's unneeded and perform the testing logic separately. I believe I've accomplished this but if it appears I've missed anything it's not out of the question to leave it as-is for now.

Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStoreTest.java:313]: You raise good points about how the existing checks can be done on the result of the method invocation, and since we've mocked the CloseableOffsetStorageReader implementation we do know what we returned. I agree that in general this seemed like a complication that was only used in testing, and that we have not lost any visibility within the tests. Thanks for the quick response.

**Review by rhauch — APPROVED** [review]
LGTM. Thanks, @C0urante!

Waiting for a green build before merging.

**Pull request merged**
